torch/lib/c10d/ProcessGroupGloo.cpp

```diff
@@ -322,46 +362,78 @@ std::shared_ptr<::gloo::transport::Device> ProcessGroupGloo::
 std::shared_ptr<::gloo::transport::Device> ProcessGroupGloo::
     createDeviceForHostname(const std::string& hostname) {
   ::gloo::transport::tcp::attr attr;
+  attr.hostname = hostname;
+  TORCH_CHECK(
+      doesHostnameResolveToUsableAddress(attr.hostname),
+      "Cannot resolve ",
+      hostname,
+      " to a (local) address");
+  return ::gloo::transport::tcp::CreateDevice(attr);
+}
+#endif
 
-  if (hostname.empty()) {
-    // Use the hostname to resolve the network address to
-    // use. Note: if the hostname does not resolve to an address (e.g.
-    // because of misconfigured /etc/hosts file), this will not work.
-    std::array<char, HOST_NAME_MAX> buffer{};
-    auto rv = gethostname(buffer.data(), buffer.size());
-    if (rv != 0) {
-      throw std::system_error(errno, std::system_category());
-    }
-    attr.hostname = buffer.data();
-  } else {
-    attr.hostname = hostname;
+#ifdef __APPLE__
+std::shared_ptr<::gloo::transport::Device> ProcessGroupGloo::
+    createDeviceForHostname(const std::string& hostname) {
+  ::gloo::transport::uv::attr attr;
+  attr.hostname = hostname;
+  TORCH_CHECK(
+      doesHostnameResolveToUsableAddress(attr.hostname),
+      "Cannot resolve ",
+      hostname,
+      " to a (local) address");
+  return ::gloo::transport::uv::CreateDevice(attr);
+}
+#endif
+
+#ifdef __linux__
+std::shared_ptr<::gloo::transport::Device> ProcessGroupGloo::
+    createDefaultDevice() {
+  ::gloo::transport::tcp::attr attr;
+
+  // Use the hostname to resolve the network address to
+  // use. Note: if the hostname does not resolve to an address (e.g.
+  // because of misconfigured /etc/hosts file), this will not work.
+  std::array<char, HOST_NAME_MAX> buffer{};
+  auto rv = gethostname(buffer.data(), buffer.size());
+  if (rv != 0) {
+    throw std::system_error(errno, std::system_category());
   }
+  attr.hostname = buffer.data();
 
-  return ::gloo::transport::tcp::CreateDevice(attr);
+  // Use this machine's hostname if it resolves to an address.
+  if (doesHostnameResolveToUsableAddress(attr.hostname)) {
+    return ::gloo::transport::tcp::CreateDevice(attr);
+  }
+
+  // Otherwise, use the loopback interface.
+  return createDeviceForInterface("lo");
```

Comment:
will there be any log telling users we are binding to the loopback or resolved hostname address?

Reply:
There isn't now, but there should be one indeed.